src/main/java/pl/sdacademy/carrental/services/BranchService.java

```diff
@@ -47,34 +52,75 @@ public void createBranch(final BranchForm branchForm) {
     }
 
     public void delete(final Long id) {
-        final Branch branch = branchRepository.findById(id).orElseThrow();
+
+        final Branch branch = findExistingById(id);
+        final Address address = branch.getAddress();
+
+        throwIfBranchHasCarsOrEmployees(branch);
+
         branchRepository.delete(branch);
+        addressRepository.delete(address);
+    }
+
+    private Branch findExistingById(final Long id) {
+        return branchRepository.findById(id).orElseThrow(() ->
+                new BranchException("No branch with id " + id + " found."));
+    }
+
+    private void throwIfBranchHasCarsOrEmployees(final Branch branch) {
+        boolean isDeletable = true;
+        String expMessage = "";
+        if (!branch.getCarsOnHand().isEmpty()) {
+            isDeletable = false;
+            expMessage += ("Cannot remove a branch with assigned car(s). ");
+        }
+        if (!branch.getEmployees().isEmpty()) {
+            isDeletable = false;
+            expMessage += ("Cannot remove a branch with assigned employee(s).");
+        }
+        if (!isDeletable) {
+            throw new BranchException(expMessage);
+        }
     }
 
     public void updateBranch(final Long id, final BranchForm branchForm) {
 
         final Address address = createAddressFromBranchForm(branchForm);
+        addressRepository.save(address);
 
-        final Branch branch = branchRepository.findById(id).orElseThrow();
+        final Branch branch = findExistingById(id);
         branch.setAddress(address);
         branch.setName(branchForm.getName());
         branch.setStatus(branchForm.getStatus());
         branchRepository.save(branch);
     }
 
-    private Address createAddressFromBranchForm(final BranchForm branchForm){
-        final Address address = Address.builder()
+    private Address createAddressFromBranchForm(final BranchForm branchForm) {
+        return Address.builder()
                 .city(branchForm.getCity())
                 .street(branchForm.getStreet())
                 .building(branchForm.getBuilding())
                 .apartment(branchForm.getApartment())
                 .zipCode(branchForm.getZipCode())
                 .build();
-        return addressRepository.save(address);
+    }
+
+    private void throwIfBranchWithGivenAddressExists(final Address address) {
+
+        addressRepository.findAll().stream()
+                .filter(adrs ->
+                        adrs.getCity().equals(address.getCity())
+                                && adrs.getZipCode().equals(address.getZipCode())
```

Comment:
to wyglada jakby powinna byc uzyta metoda equals :)

Reply:
Hmmm, ale one mają inne id, więc equals tu zwróci zawsze false IMO.

Reply:
A nie byłoby lepiej strzelić takie zapytanie do bazy, niż pobierać całą tabelę i streamować?

Reply:
No w sumie tak by było znacznie lepiej :)